turbo/engineapi/request_list.go

```diff
@@ -143,34 +151,48 @@ func (rl *RequestList) WaitForRequest(onlyNew bool, noWait bool) (interrupt Inte
 }
 
 func (rl *RequestList) IsWaiting() bool {
-	return atomic.LoadUint32(&rl.waiting) != 0
+	return rl.waiting.Load() != 0
 }
 
 func (rl *RequestList) updateWaiting(val uint32) {
-	rl.waiterMtx.Lock()
-	defer rl.waiterMtx.Unlock()
-	atomic.StoreUint32(&rl.waiting, val)
+	rl.waiting.Store(val)
 
 	if val == 1 {
-		// something might be waiting to be notified of the waiting state being ready
-		for _, c := range rl.waiters {
-			c <- struct{}{}
-		}
-		rl.waiters = make([]chan struct{}, 0)
+		rl.notifyWaiters()
 	}
 }
 
-func (rl *RequestList) WaitForWaiting(c chan struct{}) {
+func (rl *RequestList) WaitForWaiting(ctx context.Context) {
 	rl.waiterMtx.Lock()
-	defer rl.waiterMtx.Unlock()
-	val := atomic.LoadUint32(&rl.waiting)
-	if val == 1 {
-		// we are already waiting so just send to the channel and quit
-		c <- struct{}{}
+	isWaiting := rl.waiting.Load()
+	if isWaiting == 1 {
+		// we are already waiting so just return
+		rl.waiterMtx.Unlock()
+		return
 	} else {
-		// we need to register a waiter now to be notified when we are ready
+		c := make(chan struct{}, 1)
 		rl.waiters = append(rl.waiters, c)
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			// wait for either the context to be done or the waiter to be triggered
+			select {
+			case <-c:
+				wg.Done()
+			case <-ctx.Done():
+				wg.Done()
+			}
+		}()
+		rl.waiterMtx.Unlock()
+		wg.Wait()
+	}
+}
+
+func (rl *RequestList) notifyWaiters() {
+	for _, c := range rl.waiters {
+		close(c)
 	}
+	rl.waiters = make([]chan struct{}, 0)
```

Comment:
this is not thread-safe

Reply:
the function is only ever called within a locked context, but it's not all that clear that this is the case.  You mentioned the other day about only public functions managing mutex locks to avoid double locking scenarios.  Is this still desired?  If not I can move the locks closer to this code

Reply:
then it's ok